File: dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/State/TurnState.cs
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;
using Microsoft.Teams.AI.Exceptions;
using Microsoft.Teams.AI.Utilities;

namespace Microsoft.Teams.AI.State
{
    /// <summary>
    /// Base class defining a collection of turn state scopes.
    /// Developers can create a derived class that extends `TurnState` to add additional state scopes.
    /// </summary>
    public class TurnState : IMemory
    {
        private Dictionary<string, TurnStateEntry> _scopes = new();
        private bool _isLoaded = false;
        private Task<bool>? _loadingTask = Task.FromResult(false);

        /// <summary>
        /// Name of the conversation scope.
        /// </summary>
        public const string CONVERSATION_SCOPE = "conversation";

        /// <summary>
        /// Name of the user scope.
        /// </summary>
        public const string USER_SCOPE = "user";

        /// <summary>
        /// Name of the temp scope.
        /// </summary>
        public const string TEMP_SCOPE = "temp";

        /// <summary>
        /// Initializes a new instance of the <see cref="TurnState"/> class.
        /// </summary>
        public TurnState()
        {
            ScopeDefaults = new Dictionary<string, Record>();
            ScopeDefaults.Add(CONVERSATION_SCOPE, new Record());
            ScopeDefaults.Add(USER_SCOPE, new Record());
        }

        /// <summary>
        /// The default values to initial for each scope.
        /// </summary>
        protected Dictionary<string, Record> ScopeDefaults;

        /// <summary>
        /// Provides the current status of the load.
        /// </summary>
        /// <returns>Returns true if the scopes have been loaded, false otherwise.</returns>
        public bool IsLoaded() { return _isLoaded; }

        /// <summary>
        /// Returns the TurnStateEntry associated with the specified scope.
        /// </summary>
        /// <param name="name">The specified scope.</param>
        /// <returns>The state saved for the scope.</returns>
        public TurnStateEntry? GetScope(string name)
        {
            try
            {
                return _scopes[name];
            }
            catch (KeyNotFoundException)
            {
                return null;
            }
        }

        /// <summary>
        /// Stores all the conversation-related state.
        /// </summary>
        public Record Conversation
        {
            get
            {
                TurnStateEntry? scope = GetScope(CONVERSATION_SCOPE);
                if (scope == null)
                {
                    throw new ArgumentException("TurnState hasn't been loaded. Call LoadStateAsync() first.");
                }

                return scope.Value!;
            }
            set
            {
                Verify.ParamNotNull(value);

                TurnStateEntry? scope = GetScope(CONVERSATION_SCOPE);
                if (scope == null)
                {
                    throw new ArgumentException("TurnState hasn't been loaded. Call LoadStateAsync() first.");
                }

                scope.Replace(value!);
            }
        }

        /// <summary>
        /// Stores all the user related state.
        /// </summary>
        public Record User
        {
            get
            {
                TurnStateEntry? scope = GetScope(USER_SCOPE);
                if (scope == null)
                {
                    throw new ArgumentException("TurnState hasn't been loaded. Call LoadStateAsync() first.");
                }

                return scope.Value!;
            }
            set
            {
                Verify.ParamNotNull(value);

                TurnStateEntry? scope = GetScope(USER_SCOPE);
                if (scope == null)
                {
                    throw new ArgumentException("TurnState hasn't been loaded. Call LoadStateAsync() first.");
                }

                scope.Replace(value!);
            }
        }

        /// <summary>
        /// Stores all the temporary state for the current turn.
        /// </summary>
        public TempState Temp
        {
            get
            {
                TurnStateEntry? scope = GetScope(TEMP_SCOPE);
                if (scope == null)
                {
                    throw new ArgumentException("TurnState hasn't been loaded. Call LoadStateAsync() first.");
                }

                return (TempState)scope.Value!;
            }
            set
            {
                Verify.ParamNotNull(value);

                TurnStateEntry? scope = GetScope(TEMP_SCOPE);
                if (scope == null)
                {
                    throw new ArgumentException("TurnState hasn't been loaded. Call LoadStateAsync() first.");
                }

                scope.Replace(value!);
            }
        }

        /// <summary>
        /// Deletes the conversation state.
        /// </summary>
        public void DeleteConversationState()
        {
            TurnStateEntry? scope = GetScope(CONVERSATION_SCOPE);
            if (scope == null)
            {
                throw new ArgumentException("TurnState hasn't been loaded. Call LoadStateAsync() first.");
            }

            scope.Delete();
        }

        /// <summary>
        /// Deletes the temporary state.
        /// </summary>
        public void DeleteTempState()
        {
            TurnStateEntry? scope = GetScope(TEMP_SCOPE);
            if (scope == null)
            {
                throw new ArgumentException("TurnState hasn't been loaded. Call LoadStateAsync() first.");
            }
            scope.Delete();
        }

        /// <summary>
        /// Deletes the user state
        /// </summary>
        public void DeleteUserState()
        {
            TurnStateEntry? scope = GetScope(USER_SCOPE);
            if (scope == null)
            {
                throw new ArgumentException("TurnState hasn't been loaded. Call LoadStateAsync() first.");
            }
            scope.Delete();
        }

        /// <summary>
        /// Deletes a value from the memory.
        /// </summary>
        /// <param name="path">Path to the value to delete in the form of `[scope].property`.
        /// If scope is omitted, the value is deleted from the temporary scope.</param>
        public void DeleteValue(string path)
        {
            (TurnStateEntry scope, string name) = GetScopeAndName(path);
            if (scope.Value!.ContainsKey(name))
            {
                scope.Value.Remove(name);
            }
        }

        /// <summary>
        /// Checks if a value exists in the memory.
        /// </summary>
        /// <param name="path"> Path to the value to check in the form of `[scope].property`.
        /// If scope is omitted, the value is checked in the temporary scope.</param>
        /// <returns>True if the value exists, false otherwise.</returns>
        public bool HasValue(string path)
        {
            (TurnStateEntry scope, string name) = GetScopeAndName(path);
            return scope.Value?.ContainsKey(name) == true;
        }

        /// <summary>
        /// Retrieves a value from the memory.
        /// </summary>
        /// <param name="path">Path to the value to retrieve in the form of `[scope].property`.
        /// If scope is omitted, the value is retrieved from the temporary scope.</param>
        /// <returns>The value or undefined if not found.</returns>
        public object? GetValue(string path)
        {
            (TurnStateEntry scope, string name) = GetScopeAndName(path);

            if (scope.Value?.ContainsKey(name) != true)
            {
                return null;
            }

            return scope.Value[name];
        }

        /// <summary>
        /// Assigns a value to the memory.
        /// </summary>
        /// <param name="path">Path to the value to assign in the form of `[scope].property`.
        /// If scope is omitted, the value is assigned to the temporary scope.</param>
        /// <param name="value">Value to assign.</param>
        public void SetValue(string path, object value)
        {
            (TurnStateEntry scope, string name) = GetScopeAndName(path);
            scope.Value![name] = value;
        }

        /// <summary>
        /// Loads all of the state scopes for the current turn.
        /// </summary>
        /// <param name="storage">Optional. Storage provider to load state scopes from.</param>
        /// <param name="turnContext">Context for the current turn of conversation with the user.</param>
        /// <param name="cancellationToken">Optional. The cancellation token.</param>   
        /// <returns>True if the states need to be loaded.</returns>
        public async Task<bool> LoadStateAsync(IStorage? storage, ITurnContext turnContext, CancellationToken cancellationToken = default)
        {
            // Only load on first call
            if (this._isLoaded)
            {
                return false;
            }

            // Check for existing load operation
            if (!(await this._loadingTask!))
            {
                this._loadingTask = Task.Run(async () =>
                {
                    try
                    {
                        // Prevent additional load attempts
                        this._isLoaded = true;

                        // Compute state keys
                        List<string> keys = new();
                        Dictionary<string, string> scopes = OnComputeStorageKeys(turnContext);
                        foreach (KeyValuePair<string, string> scope in scopes)
                        {
                            if (scopes.ContainsKey(scope.Key))
                            {
                                keys.Add(scopes[scope.Key]);
                            }
                        }

                        // Read items from storage provider (if configured)
                        IDictionary<string, object> items;
                        if (storage != null)
                        {
                            items = await storage.ReadAsync(keys.ToArray(), cancellationToken);
                        }
                        else
                        {
                            items = new Dictionary<string, object>();
                        }

                        // Create scopes for items
                        foreach (KeyValuePair<string, string> scope in scopes)
                        {
                            if (scopes.ContainsKey(scope.Key))
                            {
                                Record scopeDefault = ScopeDefaults.ContainsKey(scope.Key) ? ScopeDefaults[scope.Key] : new Record();
                                string storageKey = scopes[scope.Key];
                                object value = items.ContainsKey(storageKey) ? items[storageKey] : scopeDefault;
                                this._scopes[scope.Key] = new TurnStateEntry((value as Record)!, storageKey);
                            }
                        }

                        // Add the temp scope
                        this._scopes[TEMP_SCOPE] = new TurnStateEntry(new TempState());

                        // Clear loading task
                        this._isLoaded = true;
                        this._loadingTask = null;
                        return true;
                    }
                    catch (Exception ex)
                    {
                        this._loadingTask = null;
                        throw new TeamsAIException($"Something went wrong when loading state: {ex.Message}", ex);
                    }
                });
            }

            return this._loadingTask.Result;
        }

        /// <summary>
        /// Saves all of the state scopes for the current turn.
        /// </summary>
        /// <param name="turnContext">Context for the current turn of conversation with the user.</param>
        /// <param name="storage">Optional. Storage provider to save state scopes to.</param>
        public async Task SaveStateAsync(ITurnContext turnContext, IStorage? storage)
        {
            Verify.ParamNotNull(turnContext);

            // Check for existing load operation
            if (!this._isLoaded && this._loadingTask!.Result)
            {
                // Wait for load to finish
                await this._loadingTask;
            }

            // Ensure loaded
            if (!this._isLoaded)
            {
                throw new ArgumentNullException($"TurnState hasn't been loaded. Call loadState() first.");
            }

            // Find changes and deletions
            Record changes = new();
            List<string> deletions = new();

            foreach (KeyValuePair<string, TurnStateEntry> scope in this._scopes)
            {
                if (!this._scopes.ContainsKey(scope.Key))
                {
                    continue;
                }
                TurnStateEntry entry = this._scopes[scope.Key];
                if (entry.StorageKey != null)
                {
                    if (entry.IsDeleted)
                    {
                        // Add to deletion list
                        if (deletions != null)
                        {
                            deletions.Add(entry.StorageKey);
                        }
                        else
                        {
                            deletions = new() { entry.StorageKey };
                        }
                    }
                    else if (entry.HasChanged)
                    {
                        // Add to change set
                        if (changes == null)
                        {
                            changes = new();
                        }
                        changes[entry.StorageKey] = entry.Value!;
                    }
                }
            }

            // Do we have a storage provider?
            if (storage != null)
            {
                // Apply changes
                List<Task> tasks = new();
                if (changes.Keys.Count > 0)
                {
                    tasks.Add(storage.WriteAsync(changes));
                }

                // Apply deletions
                if (deletions.Count > 0)
                {
                    tasks.Add(storage.DeleteAsync(deletions.ToArray()));
                }

                if (tasks.Count > 0)
                {
                    await Task.WhenAll(tasks.ToArray());
                }
            }
        }

        /// <summary>
        /// Computes the state keys
        /// </summary>
        /// <param name="context">Context for the current turn.</param>
        /// <returns>Stored conversation and user scopes.</returns>
        protected virtual Dictionary<string, string> OnComputeStorageKeys(ITurnContext context)
        {
            // Compute state keys
            Activity activity = context.Activity;
            string channelId = activity.ChannelId;
            string botId = activity.Recipient.Id;
            string conversationId = activity.Conversation.Id;
            string userId = activity.From.Id;

            Verify.ParamNotNull(activity, "TurnContext.Activity");
            Verify.ParamNotNull(channelId, "TurnContext.Activity.ChannelId");
            Verify.ParamNotNull(botId, "TurnContext.Activity.Recipient.Id");
            Verify.ParamNotNull(conversationId, "TurnContext.Activity.Conversation.Id");
            Verify.ParamNotNull(userId, "TurnContext.Activity.From.Id");

            string conversationKey = $"{channelId}/${botId}/conversations/${conversationId}";
            string userKey = $"{channelId}/${botId}/users/${userId}";

            Dictionary<string, string> keys = new();
            keys.Add(CONVERSATION_SCOPE, conversationKey);
            keys.Add(USER_SCOPE, userKey);
            return keys;
        }

        private (TurnStateEntry, string) GetScopeAndName(string path)
        {
            // Get variable scope and name
            string[] parts = path.Split('.');
            if (parts.Length > 2)
            {
                throw new ArgumentException($"Invalid state path: ${path}");
            }
            else if (parts.Length == 1)
            {
                parts.Prepend(TEMP_SCOPE);
            }

            // Validate scope
            TurnStateEntry? scope = GetScope(parts[0]);
            if (scope == null)
            {
                throw new ArgumentNullException($"Invalid state scope: ${parts[0]}");
            }

            return (scope, parts[1]);
        }
    }
}

File: js/packages/teams-ai/src/TurnState.ts
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { TurnContext, Storage, StoreItems } from 'botbuilder';
import { Memory } from './MemoryFork';
import { InputFile } from './InputFileDownloader';

/**
 * @private
 */
const CONVERSATION_SCOPE = 'conversation';

/**
 * @private
 */
const USER_SCOPE = 'user';

/**
 * @private
 */
const TEMP_SCOPE = 'temp';

/**
 * Default conversation state
 * @remarks
 * Inherit a new interface from this base interface to strongly type the applications conversation
 * state.
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface DefaultConversationState {}

/**
 * Default user state
 * @remarks
 * Inherit a new interface from this base interface to strongly type the applications user
 * state.
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface DefaultUserState {}

/**
 * Default temp state
 * @remarks
 * Inherit a new interface from this base interface to strongly type the applications temp
 * state.
 */
export interface DefaultTempState {
    /**
     * Input passed from the user to the AI Library
     */
    input: string;

    /**
     * Downloaded files passed by the user to the AI Library
     */
    inputFiles: InputFile[];

    /**
     * Output returned from the last executed action
     */
    lastOutput: string;

    /**
     * All outputs returned from the action sequence that was executed
     */
    actionOutputs: Record<string, string>;

    /**
     * User authentication tokens
     */
    authTokens: { [key: string]: string };

    /**
     * Flag indicating whether a token exchange event has already been processed
     */
    duplicateTokenExchange?: boolean;
}

/**
 * Base class defining a collection of turn state scopes.
 * @remarks
 * Developers can create a derived class that extends `TurnState` to add additional state scopes.
 * ```JavaScript
 * class MyTurnState extends TurnState {
 *   protected async onComputeStorageKeys(context) {
 *     const keys = await super.onComputeStorageKeys(context);
 *     keys['myScope'] = `myScopeKey`;
 *     return keys;
 *   }
 *
 *   public get myScope() {
 *     const scope = this.getScope('myScope');
 *     if (!scope) {
 *       throw new Error(`MyTurnState hasn't been loaded. Call load() first.`);
 *     }
 *     return scope.value;
 *   }
 *
 *   public set myScope(value) {
 *     const scope = this.getScope('myScope');
 *     if (!scope) {
 *       throw new Error(`MyTurnState hasn't been loaded. Call load() first.`);
 *     }
 *     scope.replace(value);
 *   }
 * }
 * ```
 */
/**
 * Represents the turn state for a conversation.
 * Turn state includes conversation state, user state, and temporary state.
 * Provides methods to access, modify, and delete the state objects.
 * @template TConversationState
 * @param {TConversationState} TConversationState The type of the conversation state object.
 * @param {TUserState} TUserState The type of the user state object.
 * @param {TTempState} TTempState - The type of the temporary state object.
 */
export class TurnState<
    TConversationState = DefaultConversationState,
    TUserState = DefaultUserState,
    TTempState = DefaultTempState
> implements Memory
{
    private _scopes: Record<string, TurnStateEntry> = {};
    private _isLoaded = false;
    private _loadingPromise?: Promise<boolean>;
    private _stateNotLoadedString = `TurnState hasn't been loaded. Call load() first.`;

    /**
     * Accessor for the conversation state.
     */
    /**
     * Gets the conversation state from the turn state.
     * @template TConversationState
     * @returns {TConversationState} The conversation state.
     * @throws Error if TurnState hasn't been loaded. Call load() first.
     */
    public get conversation(): TConversationState {
        const scope = this.getScope(CONVERSATION_SCOPE);
        if (!scope) {
            throw new Error(this._stateNotLoadedString);
        }
        return scope.value as TConversationState;
    }

    /**
     * Replaces the conversation state with a new value.
     * @param {TConversationState} value New value to replace the conversation state with.
     */
    public set conversation(value: TConversationState) {
        const scope = this.getScope(CONVERSATION_SCOPE);
        if (!scope) {
            throw new Error(this._stateNotLoadedString);
        }
        scope.replace(value as Record<string, unknown>);
    }

    /**
     * Gets a value indicating whether the applications turn state has been loaded.
     * @returns {boolean} True if the applications turn state has been loaded, false otherwise.
     */
    public get isLoaded(): boolean {
        return this._isLoaded;
    }

    /**
     * Accessor for the temp state.
     @returns {TTempState} The temp TurnState.
     @throws Error if TurnState hasn't been loaded. Call load() first.
     */
    public get temp(): TTempState {
        const scope = this.getScope(TEMP_SCOPE);
        if (!scope) {
            throw new Error(this._stateNotLoadedString);
        }
        return scope.value as TTempState;
    }

    /**
     * Replaces the temp state with a new value.
     * @param {TTempState} value New value to replace the temp state with.
     * @throws Error if TurnState hasn't been loaded. Call load() first.
     */
    public set temp(value: TTempState) {
        const scope = this.getScope(TEMP_SCOPE);
        if (!scope) {
            throw new Error(this._stateNotLoadedString);
        }
        scope.replace(value as Record<string, unknown>);
    }

    /**
     * Accessor for the user state.
     * @returns {TUserState} The user TurnState.
     */
    public get user(): TUserState {
        const scope = this.getScope(USER_SCOPE);
        if (!scope) {
            throw new Error(this._stateNotLoadedString);
        }
        return scope.value as TUserState;
    }

    /**
     * Replaces the user state with a new value.
     */
    public set user(value: TUserState) {
        const scope = this.getScope(USER_SCOPE);
        if (!scope) {
            throw new Error(this._stateNotLoadedString);
        }
        scope.replace(value as Record<string, unknown>);
    }

    /**
     * Deletes the state object for the current conversation from storage.
     */
    public deleteConversationState(): void {
        const scope = this.getScope(CONVERSATION_SCOPE);
        if (!scope) {
            throw new Error(this._stateNotLoadedString);
        }
        scope.delete();
    }

    /**
     * Deletes the temp state object.
     */
    public deleteTempState(): void {
        const scope = this.getScope(TEMP_SCOPE);
        if (!scope) {
            throw new Error(this._stateNotLoadedString);
        }
        scope.delete();
    }

    /**
     * Deletes the state object for the current user from storage.
     */
    public deleteUserState(): void {
        const scope = this.getScope(USER_SCOPE);
        if (!scope) {
            throw new Error();
        }
        scope.delete();
    }

    /**
     * Gets a state scope by name.
     * @param {string} scope Name of the state scope to return. (i.e. 'conversation', 'user', or 'temp')
     * @returns {string | undefined} The state scope or undefined if not found.
     */
    public getScope(scope: string): TurnStateEntry | undefined {
        return this._scopes[scope];
    }

    /**
     * Deletes a value from the memory.
     * @param {string} path Path to the value to delete in the form of `[scope].property`. If scope is omitted, the value is deleted from the temporary scope.
     */
    public deleteValue(path: string): void {
        const { scope, name } = this.getScopeAndName(path);
        if (scope.value.hasOwnProperty(name)) {
            delete scope.value[name];
        }
    }

    /**
     * Checks if a value exists in the memory.
     * @param {string} path Path to the value to check in the form of `[scope].property`. If scope is omitted, the value is checked in the temporary scope.
     * @returns {boolean} True if the value exists, false otherwise.
     */
    public hasValue(path: string): boolean {
        const { scope, name } = this.getScopeAndName(path);
        return Object.prototype.hasOwnProperty.call(scope.value, name);
    }

    /**
     * Retrieves a value from the memory.
     * @template TValue
     * @param {TValue} path Path to the value to retrieve in the form of `[scope].property`. If scope is omitted, the value is retrieved from the temporary scope.
     * @returns {string} The value or undefined if not found.
     */
    public getValue<TValue = unknown>(path: string): TValue {
        const { scope, name } = this.getScopeAndName(path);
        return scope.value[name] as TValue;
    }

    /**
     * Assigns a value to the memory.
     * @param {string} path Path to the value to assign in the form of `[scope].property`. If scope is omitted, the value is assigned to the temporary scope.
     * @param {unknown} value Value to assign.
     */
    public setValue(path: string, value: unknown): void {
        const { scope, name } = this.getScopeAndName(path);
        scope.value[name] = value;
    }

    /**
     * Loads all of the state scopes for the current turn.
     * @param {TurnContext} context Context for the current turn of conversation with the user.
     * @param {Storage} storage Optional. Storage provider to load state scopes from.
     * @returns {boolean} True if the states needed to be loaded.
     */
    public load(context: TurnContext, storage?: Storage): Promise<boolean> {
        // Only load on first call
        if (this._isLoaded) {
            return Promise.resolve(false);
        }

        // Check for existing load operation
        if (!this._loadingPromise) {
            this._loadingPromise = new Promise<boolean>(async (resolve, reject) => {
                try {
                    // Prevent additional load attempts
                    this._isLoaded = true;

                    // Compute state keys
                    const keys: string[] = [];
                    const scopes = await this.onComputeStorageKeys(context);
                    for (const key in scopes) {
                        if (Object.prototype.hasOwnProperty.call(scopes, key)) {
                            keys.push(scopes[key]);
                        }
                    }

                    // Read items from storage provider (if configured)
                    const items = storage ? await storage.read(keys) : {};

                    // Create scopes for items
                    for (const key in scopes) {
                        if (Object.prototype.hasOwnProperty.call(scopes, key)) {
                            const storageKey = scopes[key];
                            const value = items[storageKey];
                            this._scopes[key] = new TurnStateEntry(value, storageKey);
                        }
                    }

                    // Add the temp scope
                    this._scopes[TEMP_SCOPE] = new TurnStateEntry({});

                    // Clear loading promise
                    this._isLoaded = true;
                    this._loadingPromise = undefined;
                    resolve(true);
                } catch (err) {
                    this._loadingPromise = undefined;
                    reject(err);
                }
            });
        }

        return this._loadingPromise;
    }

    /**
     * Saves all of the state scopes for the current turn.
     * @param {TurnContext} context Context for the current turn of conversation with the user.
     * @param {Storage} storage Optional. Storage provider to save state scopes to.
     */
    public async save(context: TurnContext, storage?: Storage): Promise<void> {
        // Check for existing load operation
        if (!this._isLoaded && this._loadingPromise) {
            // Wait for load to finish
            await this._loadingPromise;
        }

        // Ensure loaded
        if (!this._isLoaded) {
            throw new Error(this._stateNotLoadedString);
        }

        // Find changes and deletions
        let changes: StoreItems | undefined;
        let deletions: string[] | undefined;
        for (const key in this._scopes) {
            if (!Object.prototype.hasOwnProperty.call(this._scopes, key)) {
                continue;
            }
            const entry = this._scopes[key];
            if (entry.storageKey) {
                if (entry.isDeleted) {
                    // Add to deletion list
                    if (deletions) {
                        deletions.push(entry.storageKey);
                    } else {
                        deletions = [entry.storageKey];
                    }
                } else if (entry.hasChanged) {
                    // Add to change set
                    if (!changes) {
                        changes = {};
                    }

                    changes[entry.storageKey] = entry.value;
                }
            }
        }

        // Do we have a storage provider?
        if (storage) {
            // Apply changes
            const promises: Promise<void>[] = [];
            if (changes) {
                promises.push(storage.write(changes));
            }

            // Apply deletions
            if (deletions) {
                promises.push(storage.delete(deletions));
            }

            // Wait for completion
            if (promises.length > 0) {
                await Promise.all(promises);
            }
        }
    }

    /**
     * Computes the storage keys for the state scopes being persisted.
     * @remarks
     * Can be overridden in derived classes to add additional storage scopes.
     * @param {TurnContext} context Context for the current turn of conversation with the user.
     * @returns {Promise<Record<string, string>>} A dictionary of scope names -> storage keys.
     * @throws Error if the context is missing a required property.
     */
    protected onComputeStorageKeys(context: TurnContext): Promise<Record<string, string>> {
        // Compute state keys
        const activity = context.activity;
        const channelId = activity?.channelId;
        const botId = activity?.recipient?.id;
        const conversationId = activity?.conversation?.id;
        const userId = activity?.from?.id;

        if (!channelId) {
            throw new Error('missing context.activity.channelId');
        }

        if (!botId) {
            throw new Error('missing context.activity.recipient.id');
        }

        if (!conversationId) {
            throw new Error('missing context.activity.conversation.id');
        }

        if (!userId) {
            throw new Error('missing context.activity.from.id');
        }

        const keys: Record<string, string> = {};
        keys[CONVERSATION_SCOPE] = `${channelId}/${botId}/conversations/${conversationId}`;
        keys[USER_SCOPE] = `${channelId}/${botId}/users/${userId}`;
        return Promise.resolve(keys);
    }

    /**
     * @private
     * @param {string} path Path to the value to check in the form of `[scope].property`. If scope is omitted, the value is checked in the temporary scope.
     * @returns {{ scope: TurnStateEntry; name: string }} Scope and name.
     */
    private getScopeAndName(path: string): { scope: TurnStateEntry; name: string } {
        // Get variable scope and name
        const parts = path.split('.');
        if (parts.length > 2) {
            throw new Error(`Invalid state path: ${path}`);
        } else if (parts.length === 1) {
            parts.unshift(TEMP_SCOPE);
        }

        // Validate scope
        const scope = this.getScope(parts[0]);
        if (scope === undefined) {
            throw new Error(`Invalid state scope: ${parts[0]}`);
        }
        return { scope, name: parts[1] };
    }
}

/**
 * Accessor class for managing an individual state scope.
 */
export class TurnStateEntry {
    private _value: Record<string, unknown>;
    private _storageKey?: string;
    private _deleted = false;
    private _hash: string;

    /**
     * Creates a new instance of the `TurnStateEntry` class.
     * @param {Record<string, unknown>} value Optional. Value to initialize the state scope with. The default is an {} object.
     * @param {string} storageKey Optional. Storage key to use when persisting the state scope.
     */
    public constructor(value?: Record<string, unknown>, storageKey?: string) {
        this._value = value || {};
        this._storageKey = storageKey;
        this._hash = JSON.stringify(this._value);
    }

    /**
     * Gets a value indicating whether the state scope has changed since it was last loaded.
     * @returns {boolean} A value indicating whether the state scope has changed.
     */
    public get hasChanged(): boolean {
        return JSON.stringify(this._value) != this._hash;
    }

    /**
     * Gets a value indicating whether the state scope has been deleted.
     * @returns {boolean} A value indicating whether the state scope has been deleted.
     */
    public get isDeleted(): boolean {
        return this._deleted;
    }

    /**
     * Gets the value of the state scope.
     * @returns {Record<string, unknown>} value of the state scope.
     */
    public get value(): Record<string, unknown> {
        if (this.isDeleted) {
            // Switch to a replace scenario
            this._value = {};
            this._deleted = false;
        }

        return this._value;
    }

    /**
     * Gets the storage key used to persist the state scope.
     * @returns {string | undefined} The storage key used to persist the state scope.
     */
    public get storageKey(): string | undefined {
        return this._storageKey;
    }

    /**
     * Clears the state scope.
     */
    public delete(): void {
        this._deleted = true;
    }

    /**
     * Replaces the state scope with a new value.
     * @template TValue
     * @param {TValue} value New value to replace the state scope with.
     */
    public replace(value?: Record<string, unknown>): void {
        this._value = value || {};
    }
}
